contract/LibSubmarine.sol

```diff
@@ -43,135 +36,177 @@ contract LibSubmarine is ProvethVerifier {
         uint8 _result,
         address indexed _sender,
         address indexed _submarine,
-        uint256 _slashAmount,
+        uint256 _slashAmount/*,
         bytes _proofBlob,
-        bytes _unsignedCommitTx
+        bytes _unsignedCommitTx*/ // Removed due to stack limit
     );
     event Finalized(
         bytes32 indexed _sessionId,
         address indexed _sender,
         uint256 _amountTemp,
         uint256 _revealDeposit,
-        bytes _dataTemp
+        bytes _commitData
     );
 
+    /////////////
+    // Storage //
+    /////////////
 
-    constructor(uint256 _revealDepositAmount, uint32 _challengePeriodLength) public {
-        revealDepositAmount = _revealDepositAmount;
-        challengePeriodLength = _challengePeriodLength;
+    uint256 public revealDepositAmount;
+    uint32 public challengePeriodLength;
+    uint8 public vee = 27;
+
+    mapping(bytes32 => SessionData) public sessionData;
+
+    struct SessionData {
+        bool unlocked;        // set in unlock, used in reveal and finalize
+        uint32 revealBlock;   // set in reveal, used in unlock, challenge, and finalize
+        uint256 commitValue;  // set in reveal, used in unlock, challenge, and finalize
+        bytes32 hashedCommit; // set in reveal, used in challenge, and finalize
+    }
+
+    /* Layout of hashedCommit
```

Comment:
we should have documentation for this somewhere, rather than comments cluttering up code, methinks